File: Python/0366-m-find-leaves-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
"""
dfs
当前node的高度是 1 + Max（height(node.left), height(node.right)
leaf的高度是 0
The height of a node is also its index in the result list (res).
"""
import collections
class Solution:
    def findLeaves(self, root: TreeNode) -> List[List[int]]:
        res = []
        hash = collections.defaultdict(list)
        maxH = self.dfs(root, hash)

        for i in range(1, maxH + 1):
            res.append(hash[i])
        return res

    def dfs(self, cur, hash):
        if not cur:
            return 0

        h = max(self.dfs(cur.left, hash), self.dfs(cur.right, hash)) + 1
        hash[h].append(cur.val)

        return h

File: leetcode/C++/0366-m-find-leaves-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int dfs(TreeNode* node, unordered_map<int, vector<int>>& hash) {
        if (!node) return 0;
        int h = max(dfs(node->left, hash), dfs(node->right, hash)) + 1;
        hash[h].push_back(node->val);
        return h;
    }
    
public:
    vector<vector<int>> findLeaves(TreeNode* root) {
        unordered_map<int, vector<int>> hash;
        vector<vector<int>> res;
        int maxH = dfs(root, hash);
        for (int i = 1; i < maxH + 1; i++) 
            res.push_back(hash[i]);
        return res;
    }
};